kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h

```diff
@@ -792,90 +828,102 @@ class ResidualBasedBlockBuilderAndSolver
+        TSystemVectorType& rb
+        ) override
     {
-        std::size_t system_size = A.size1();
-        std::vector<double> scaling_factors (system_size, 0.0);
+        const std::size_t system_size = rA.size1();
+        Vector scaling_factors (system_size);
 
+        const auto it_dof_iterator_begin = BaseType::mDofSet.begin();
         const int ndofs = static_cast<int>(BaseType::mDofSet.size());
 
-        //NOTE: dofs are assumed to be numbered consecutively in the BlockBuilderAndSolver
+        // NOTE: dofs are assumed to be numbered consecutively in the BlockBuilderAndSolver
         #pragma omp parallel for firstprivate(ndofs)
         for (int k = 0; k<ndofs; k++) {
-            typename DofsArrayType::iterator dof_iterator = BaseType::mDofSet.begin() + k;
-            if(dof_iterator->IsFixed())
+            auto it_dof_iterator = it_dof_iterator_begin + k;
+            if (it_dof_iterator->IsFixed()) {
                 scaling_factors[k] = 0.0;
-            else
+            } else {
                 scaling_factors[k] = 1.0;
-
+            }
         }
 
-        double* Avalues = A.value_data().begin();
-        std::size_t* Arow_indices = A.index1_data().begin();
-        std::size_t* Acol_indices = A.index2_data().begin();
+        double* Avalues = rA.value_data().begin();
+        std::size_t* Arow_indices = rA.index1_data().begin();
+        std::size_t* Acol_indices = rA.index2_data().begin();
 
-        //detect if there is a line of all zeros and set the diagonal to a 1 if this happens
-        #pragma omp parallel for firstprivate(system_size)
-        for (int k = 0; k < static_cast<int>(system_size); ++k){
-            std::size_t col_begin = Arow_indices[k];
-            std::size_t col_end = Arow_indices[k+1];
+        // The diagonal considered
+        mScaleFactor = GetScaleNorm(rModelPart, rA);
+
+        // Detect if there is a line of all zeros and set the diagonal to a 1 if this happens
+        #pragma omp parallel firstprivate(system_size)
+        {
+            std::size_t col_begin = 0, col_end  = 0;
             bool empty = true;
-            for (std::size_t j = col_begin; j < col_end; ++j)
-            {
-                if(Avalues[j] != 0.0)
-                {
-                    empty = false;
-                    break;
+
+            #pragma omp for
+            for (int k = 0; k < static_cast<int>(system_size); ++k) {
+                col_begin = Arow_indices[k];
+                col_end = Arow_indices[k + 1];
+                empty = true;
+                for (std::size_t j = col_begin; j < col_end; ++j) {
+                    if(Avalues[j] != 0.0) {
+                        empty = false;
+                        break;
+                    }
                 }
-            }
 
-            if(empty == true)
-            {
-                A(k,k) = 1.0;
-                b[k] = 0.0;
+                if(empty) {
+                    rA(k, k) = mScaleFactor;

```

Comment:
In the scope of finalizing #6434 I checked this and the following appears strange to me:

you call it ScaleFactor but you overwrite the value... => Naming is a bit strange to me, I would have expected `rA(k, k) *= mScaleFactor`. I might be wrong

in any case, the current behavior is what I need :)